### ISSUE TITLE
w.instantiate/ error on webclient update

### ISSUE THREAD

**Issue by johnbotris** [tutao/tutanota#2989]
At every update users often receive an error when refreshing to receive the new version

```
Error
Error message: https://mail.tutanota.com/Urlifier-fe845b1b.js, (SystemJS https://git.io/JvFET#3)
Stacktrace:
w.instantiate/
```

or

```
Error
Error message: https://mail.tutanota.com/jszip-7eabad57.js, (SystemJS https://git.io/JvFET#3)
Stacktrace:
w.instantiate/
```

etc

**Comment by charlag** [comment]
This happens because of how the client is deployed. People who don't have service worker enabled will not have files cached and they will try to load files which do not exist anymore.

**Comment by johnbotris** [comment]
> People who don't have service worker enabled will not have files cached and they will try to load files which do not exist anymore.

I don't understand this - the service worker is what does the caching/reads back from the cache, is it not? So if there is no service worker then wouldn't it just be requesting all the files fresh each time?

**Comment by johnbotris** [comment]
From a user:
> Settings does not work
> 
> 
> Client: Browser
> Type: FREE
> Tutanota version: 3.84.10
> Timestamp (UTC): Wed, 07 Jul 2021 13:26:12 GMT
> User agent:
> Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:78.0) Gecko/20100101 Firefox/78.0
> Error
> Error message: https://mail.tutanota.com/settings-b45264d8.js, (SystemJS https://git.io/JvFET#3)
> Stacktrace:
> w.instantiate/

**Comment by johnbotris** [comment]
For this user it's trying to load native-main despite being in the browser??

> I was just navigating on the user settings options when the error popup appeared, I don't even know what happened wrong.
> 
> 
> Client: Browser
> Type: PREMIUM
> Tutanota version: 3.84.10
> Timestamp (UTC): Sat, 10 Jul 2021 13:28:31 GMT
> User agent:
> Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0
> Error
> Error message: https://mail.tutanota.com/native-main-27d9b1d5.js, (SystemJS https://git.io/JvFET#3)
> Stacktrace:
> w.instantiate/

**Comment by charlag** [comment]
> > People who don't have service worker enabled will not have files cached and they will try to load files which do not exist anymore.
> 
> I don't understand this - the service worker is what does the caching/reads back from the cache, is it not? So if there is no service worker then wouldn't it just be requesting all the files fresh each time?

yes but if you loaded app.js and it loads chunk-happyhash1.js and chunk-happyhash1.js does not exist then you get a poo-poo

**Comment by johnbotris** [comment]
@charlag but if serviceworker is disabled then app.js is coming from the server, so it should be up to date or?

**Comment by charlag** [comment]
1. You load app.js
2. Some time passes
3. We update server
4. You click on e.g. calendar and try to load calendar things using your old in-memory app.js
5. Those things are not there
6. Ka-boom

**Comment by johnbotris** [comment]
For this user it's trying to load native-main despite being in the browser??

> I was just navigating on the user settings options when the error popup appeared, I don't even know what happened wrong.
> 
> 
> Client: Browser
> Type: PREMIUM
> Tutanota version: 3.84.10
> Timestamp (UTC): Sat, 10 Jul 2021 13:28:31 GMT
> User agent:
> Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0
> Error
> Error message: https://mail.tutanota.com/native-main-27d9b1d5.js, (SystemJS https://git.io/JvFET#3)
> Stacktrace:
> w.instantiate/

**Comment by bedhub** [comment]
After a discussion with @armhub, @jowlo, @mpfau, @johnbotris  and @bedhub  the solution should be that we don't provide old client files from the server. Instead we should catch the error and show an message saying that the user has to reload/refresh the application. We could also use the "New version available" overlay here.

**Comment by charlag** [comment]
Seems like client update is only one of the possible issues. In most cases files are still accessible so either the request is blocked or the user is offline.

We could use systemjs hook like that:

```js
const systemCreateScript = System.constructor.prototype.createScript
System.constructor.prototype.createScript = function createScript(url) {
	const script = systemCreateScript.call(this, [url])
	script.addEventListener("error", () => {
		console.log("import error", url)
		window.location.reload()
	})
	return script
}
```

But it might be undesirable to do when just one request fails.
